infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java

```diff
@@ -151,12 +167,20 @@ private String makeUrl(ContainerPort port, String protocol, String path) {
       path = '/' + path;
     }
 
-    return protocol + "://" + hostname + ':' + port.getPublicPort() + path;
+    int serverPort = port.getPublicPort() != 0 ? port.getPublicPort() : port.getPrivatePort();
+
+    return protocol + "://" + hostname + ':' + serverPort + path;
```

Comment:
Maybe UriBuilder would be better here?

Reply:
I don't think I can use it, whereas I'm not sure whether I can use UriBuilderImpl because we don't use it in non-tests code.

Reply:
agree with Vitaliy, this code does not look nice (even if there is very small probability of changing url format in near future ) )